consensus/istanbul/core/handler_test.go

```diff
@@ -464,6 +483,236 @@ func TestCore_handlerMsg(t *testing.T) {
+		proposer        = validators.GetProposer()
+		proposerKey     = validatorKeyMap[proposer.Address()]
+		newProposal, _  = genBlockParams(lastBlock, proposerKey, 0, 1, 1)
+		malProposal, _  = genBlockParams(lastBlock, proposerKey, 0, 0, 0)
+	)
+
+	// Start istanbul core
+	istConfig := istanbul.DefaultConfig
+	istConfig.ProposerPolicy = istanbul.WeightedRandom
+	istCore := New(mockBackend, istConfig).(*core)
+	err := istCore.Start()
+	require.Nil(t, err)
+	defer istCore.Stop()
+
+	// Step 1 - Pre-prepare with correct message
+
+	// Create pre-prepare message
+	istanbulMsg, err := genIstanbulMsg(msgPreprepare, lastBlock.Hash(), newProposal, proposer.Address(), proposerKey)
+	require.Nil(t, err)
+
+	// Handle pre-prepare message
+	err = istCore.handleMsg(istanbulMsg.Payload)
+	require.Nil(t, err)
+
+	// splitSubList split current committee into benign CNs and malicious CNs
+	subList := validators.SubList(lastBlock.Hash(), istCore.currentView())
+	maliciousCNs, benignCNs := splitSubList(subList, numMalicious, proposer.Address())
+	benignCNs = append(benignCNs, proposer)
+
+	// Shortcut for sending consensus message to everyone in `CNList`
+	sendMessages := func(state uint64, proposal *types.Block, CNList []istanbul.Validator) {
+		for _, val := range CNList {
+			valKey := validatorKeyMap[val.Address()]
+			istanbulMsg, err := genIstanbulMsg(state, lastBlock.Hash(), proposal, val.Address(), valKey)
+			assert.Nil(t, err)
+			err = istCore.handleMsg(istanbulMsg.Payload)
+			// assert.Nil(t, err)
+		}
+	}
+
+	// Step 2 - Receive disagreeing prepare messages
+
+	sendMessages(msgPrepare, newProposal, benignCNs)
+	sendMessages(msgPrepare, malProposal, maliciousCNs)
+
+	if istCore.state.Cmp(StatePrepared) != 0 {
+		t.Logf("State stuck at preprepared")
+		return istCore.state
+	}
+
+	// Step 3 - Receive disagreeing commit messages
+
+	sendMessages(msgCommit, newProposal, benignCNs)
+	sendMessages(msgCommit, malProposal, maliciousCNs)
+
+	if istCore.state.Cmp(StateCommitted) != 0 {
+		// t.Logf("State stuck at prepared")
+		return istCore.state
+	} else {
+		// t.Logf("State is at committed")
+		return istCore.state
+	}
+}
+
+// TestCore_MalCN tests whether the proposer can commit when malicious CNs exist.
+func TestCore_malCN(t *testing.T) {
+	// If there are little malicious CNs, proposer can commit.
+	state := simulateMaliciousCN(t, 4, 1)
+	assert.Equal(t, StateCommitted, state)
+
+	// If there are more than 'f' malicious CNs, the proposer cannot commit, stuck at preprepared state.
+	state = simulateMaliciousCN(t, 4, 2)
+	assert.Equal(t, StatePreprepared, state)
+}
+
+// TestCore_chainSplit tests whether a chain split occurs in a certain conditions:
+// 1) the number of validators does not consist of 3f+1;
+//     e.g.) if 5 nodes, it consists of 3f+2 (f=1)
+//    This test assumes the number of validators is 5.
+// 2) the proposer is malicious; it sends two different blocks to each group
+func TestCore_chainSplit(t *testing.T) {
```

Comment:
How about making this as a parmeterized-test so that we can compare chain split case and non chain split case? It would be nice if we can test both n=5(=chain split), 7(=non chain split) cases.

Reply:
Thank you for the suggestion. reflected at [4623c59](https://github.com/klaytn/klaytn/pull/1396/commits/4623c593e9c6a13d93e2cf67664ea7d31a320b48)